docs/source/java/cdata.rst

```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
+
+**C++ Side**
+
+This application uses JNI to call Java code, but transfers data (zero-copy) via the C Data Interface instead.
+
+.. code-block:: cpp
+
+   #include <iostream>
+   #include <arrow/api.h>
+   #include <arrow/c/bridge.h>
+   #include <jni.h>
+
+   JNIEnv *CreateVM(JavaVM **jvm) {
+       JNIEnv *env;
+       JavaVMInitArgs vm_args;
+       JavaVMOption options[2];
+       options[0].optionString = "-Djava.class.path=cpptojava.jar"; // java jar name
+       options[1].optionString = "-DXcheck:jni:pedantic";
+       vm_args.version = JNI_VERSION_1_8;
+       vm_args.nOptions = 2;
+       vm_args.options = options;
+       int status = JNI_CreateJavaVM(jvm, (void **) &env, &vm_args);
+       if (status < 0) {
+           std::cout << "\n<<<<< Unable to Launch JVM >>>>>\n" << std::endl;
+           return nullptr;
+       }
+       return env;
+   }
+
+   int main() {
+       JNIEnv *env;
+       JavaVM *jvm;
+       env = CreateVM(&jvm);
+       if (env == nullptr) return EXIT_FAILURE;
+       jclass javaClassToBeCalledByCpp = env->FindClass("ToBeCalledByCpp");
+       if (javaClassToBeCalledByCpp != nullptr) {
+           jmethodID fillVector = env->GetStaticMethodID(javaClassToBeCalledByCpp,
+                                                         "fillVector",
+                                                         "(JJ)V");
+           if (fillVector != nullptr) {
+               struct ArrowSchema arrowSchema;
+               struct ArrowArray arrowArray;
+               std::cout << "\n<<<<< C++ to Java for Arrays >>>>>\n" << std::endl;
+               env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
+                                         reinterpret_cast<uintptr_t>(&arrowSchema),
+                                         reinterpret_cast<uintptr_t>(&arrowArray));
+               auto resultImportArray = arrow::ImportArray(&arrowArray, &arrowSchema);
+               std::shared_ptr<arrow::Array> array = resultImportArray.ValueOrDie();
+               std::cout << "[C++] Array: " << array->ToString() << std::endl;
+           } else {
+               std::cout << "Could not find fillVector method\n" << std::endl;
+               return EXIT_FAILURE;
+           }
+           jmethodID fillVectorSchemaRoot = env->GetStaticMethodID(javaClassToBeCalledByCpp,
+                                                                   "fillVectorSchemaRoot",
+                                                                   "(JJ)V");
+           if (fillVectorSchemaRoot != nullptr) {
+               struct ArrowSchema arrowSchema;
+               struct ArrowArray arrowArray;
+               std::cout << "\n<<<<< C++ to Java for RecordBatch >>>>>\n" << std::endl;
+               env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVectorSchemaRoot,
+                                         static_cast<long>(reinterpret_cast<uintptr_t>(&arrowSchema)),
```

Comment:
Don't we need the `static_cast<long>` above as well then?

(Also, shouldn't it be `jlong`?)

Reply:
All of these kind of declaration work well (I don't know what of this should I use?)

```cpp
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  static_cast<long>(reinterpret_cast<uintptr_t>(&arrowSchema)),
                                  static_cast<long>(reinterpret_cast<uintptr_t>(&arrowArray)));
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  reinterpret_cast<uintptr_t>(&arrowSchema),
                                  reinterpret_cast<uintptr_t>(&arrowArray));
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  (long)(&arrowSchema),
                                  (long)(&arrowArray));
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  (jlong)(&arrowSchema),
                                  (jlong)(&arrowArray));
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  (uintptr_t)(&arrowSchema),
                                  (uintptr_t)(&arrowArray));
```

Reply:
Thank you